src/metabase/sync_database/analyze.clj

```diff
@@ -134,6 +134,45 @@
         (log/debug (u/format-color 'green "Field '%s' looks like it contains valid JSON objects. Setting special_type to :type/JSON." (field/qualified-name field)))
         (assoc field-stats :special-type :type/JSON, :preview-display false)))))
 
+(defn valid-email?
+  [email]
+  (let [pattern #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?"]
+    (and (string? email) (re-matches pattern email))))
+
+
+(defn- values-are-valid-emails?
+  "`true` if at every item in VALUES is `nil` or a valid string-encoded JSON dictionary or array, and at least one of those is non-nil."
+  [values]
+  (try
+    (loop [at-least-one-non-nil-value? false, [val & more] values]
+      (cond
+        (and (not val)
+             (not (seq more))) at-least-one-non-nil-value?
+        (s/blank? val)         (recur at-least-one-non-nil-value? more)
+        ;; If val is non-nil, check that it's a JSON dictionary or array. We don't want to mark Fields containing other
+        ;; types of valid JSON values as :json (e.g. a string representation of a number or boolean)
+        :else                  (do (assert (valid-email? val))
+                                   (recur true more))))
+    (catch Throwable _
+      false)))
+
+(defn- test:email-special-type
+  "Mark FIELD as `:json` if it's textual, doesn't already have a special type, the majority of it's values are non-nil, and all of its non-nil values
+   are valid serialized JSON dictionaries or arrays."
+  [driver field field-stats]
+  (if-not (and (not (:special_type field))
+               (not (isa? (:base_type field) :type/Text)))
```

Comment:
I guess this logic is the same as the function you copied it from but it looks confusing and wrong :frowning_face: 

I think we should this one and the place you copied it from to be:

```clojure
(if (or (:special_type field)
        (not (isa? (:base_type field) :type/Text)))
  field-stats
  ;; ...
  )
```

The logic should be "check for values if it doesn't already have a special type and is textual"
but right now the logic looks like "check for values if it doesn't already have a special type and is *not* textual